dl/src/main/scala/com/intel/analytics/bigdl/models/fasterrcnn/utils/Nms.scala

```diff
@@ -0,0 +1,195 @@
+ * The goal of nms is to solve the problem that groups of several detections near the real location,
+ * ideally obtaining only one detection per object
+ */
+class Nms extends Serializable {
+
+  @transient private var areas: Tensor[Float] = _
+  @transient private var sortedScores: Tensor[Float] = _
+  @transient private var sortedInds: Tensor[Float] = _
+  @transient private var sortIndBuffer: Array[Int] = _
+  @transient private var suppressed: Array[Int] = _
+
+  private def init(size: Int): Unit = {
+    if (suppressed == null || suppressed.length < size) {
+      suppressed = new Array[Int](size)
+      sortIndBuffer = new Array[Int](size)
+    } else {
+      var i = 0
+      while (i < size) {
+        suppressed(i) = 0
+        i += 1
+      }
+    }
+    if (sortedScores == null) {
+      sortedScores = Tensor[Float]
+      sortedInds = Tensor[Float]
+      areas = Tensor[Float]
+    }
+  }
+
+  /**
+   * 1. first sort the scores from highest to lowest and get indices
+   * 2. for the bbox of first index,
+   * get the overlap between this box and the remaining bboxes
+   * put the first index to result buffer
+   * 3. update the indices by keeping those bboxes with overlap less than thresh
+   * 4. repeat 2 and 3 until the indices are empty
+   * @param x1      x1 tensor
+   * @param y1      y1 tensor
+   * @param x2      x2 tensor
+   * @param y2      y2 tensor
+   * @param scores  score tensor
+   * @param thresh  overlap thresh
+   * @param indices buffer to store indices after nms
+   * @return the length of indices after nms
+   */
+  private def nms(x1: Tensor[Float], y1: Tensor[Float], x2: Tensor[Float], y2: Tensor[Float],
+    scores: Tensor[Float], thresh: Float, indices: Array[Int]): Int = {
+    if (scores.nElement() == 0) return 0
+    require(indices.length >= scores.nElement())
+    require(x1.isContiguous())
+
+    init(scores.nElement())
+    getAreas(x1, x2, y1, y2, areas)
+    // indices start from 1
+    val orderLength = getSortedScoreInds(scores, sortIndBuffer)
+    var indexLenth = 0
+    var i = 0
+    var curInd = 0
+
+    while (i < orderLength) {
+      def nmsOne(): Unit = {
+        curInd = sortIndBuffer(i)
+        if (suppressed(curInd - 1) == 1) {
+          return
+        }
+        indices(indexLenth) = curInd
+        indexLenth += 1
+        var k = i + 1
+        while (k < orderLength) {
+          if (suppressed(sortIndBuffer(k) - 1) != 1 &&
+            isOverlapRatioGtThresh(x1, x2, y1, y2, curInd, sortIndBuffer(k), thresh)) {
+            suppressed(sortIndBuffer(k) - 1) = 1
+          }
+          k += 1
+        }
+      }
+      nmsOne()
+      i += 1
+    }
+    indexLenth
+  }
+
+  private def getSortedScoreInds(scores: Tensor[Float], resultBuffer: Array[Int]): Int = {
+    // note that when the score is the same,
+    // the order of the indices are different in python and here
+    scores.topk(scores.nElement(), dim = 1, increase = false, result = sortedScores,
```

Comment:
topk pass in a method into DimApply3. topk can be optimized special for contiguous case.